Coordinated compaction fixups

bashtanov opened the pull request:
Coordinated compaction bugfixes:
1) hold raft gate in functions submitted to single-fiber executors until their futures are resolved
2) eliminate a race condition in raft_fixture

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [x] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

bashtanov committed:
raft/compaction_coordinator: hold raft gate in submitted function

A function submitted into ssx/single_fiber_executor may be executed
late and is not held while the future returned is being awaited.
So such function should hold necessary gates until it returns a future.

bashtanov committed:
raft/tests/raft_fixture_base: eliminate a race when closing a channel

On an attempt to use an invalid channel, it was asynchronously stopped
and then removed from the map. If two fibers do this concurrently one of
them may be still co_awaiting channel::stop() when the other one has
destructed the channel object.

Remove the channel from the map synchronously while the iterator is
still valid and stop it when nothing else can access it. Concurrent
fibers will create (and potentially stop and drop) channels with the
same node id, but they will all be distinct.

This should fix unclosed-gate-destruction and unable-transfer-leadership
errors on the coordinated compaction test.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--fd4cb79f-eb40-4ebe-a104-85ef78d68762-->
<!--TEST RESULTS-->
<!--d06bb263-f1eb-4f60-b255-f503b285f7c4-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74979">build#74979</a></summary>


| test_class                        | test_method                              | test_arguments                                                               | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                | test_history                                                                                                                                                                      |
|:----------------------------------|:-----------------------------------------|:-----------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingReplicationTests     | test_auto_prefix_trimming                | {"source_cluster_spec": {"cluster_type": "redpanda"}, "with_failures": true} | integration | https://buildkite.com/redpanda/redpanda/builds/74979#019a183d-9054-4561-9726-8eadb0d79964 | FLAKY         | 17/21    | upstream reliability is '100.0'. current run reliability is '80.95238095238095'. drift is 19.04762 and the allowed drift is set to 50. The test should PASS           | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_auto_prefix_trimming             |
| VerifyConsumerOffsetsThruUpgrades | test_consumer_group_offsets              | {"versions_to_upgrade": 2}                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/74979#019a183d-9055-438e-a883-6e53ba4c64d4 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=VerifyConsumerOffsetsThruUpgrades&test_method=test_consumer_group_offsets       |
| TxAtomicProduceConsumeTest        | test_basic_tx_consumer_transform_produce | {"with_failures": true}                                                      | integration | https://buildkite.com/redpanda/redpanda/builds/74979#019a183f-225e-4888-ac38-1a3079b29ae7 | FLAKY         | 20/21    | upstream reliability is '99.37629937629939'. current run reliability is '95.23809523809523'. drift is 4.1382 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TxAtomicProduceConsumeTest&test_method=test_basic_tx_consumer_transform_produce |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75011">build#75011</a></summary>


| test_class                                     | test_method                                    | test_arguments                                                               | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                                  |
|:-----------------------------------------------|:-----------------------------------------------|:-----------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkConsumeGroupsMirroringTest           | test_continuous_group_sync                     | {"source_cluster_spec": {"cluster_type": "redpanda"}, "with_failures": true} | integration | https://buildkite.com/redpanda/redpanda/builds/75011#019a229b-0324-40e5-a484-fe1add48105d | FLAKY         | 18/21    | upstream reliability is '98.63813229571986'. current run reliability is '85.71428571428571'. drift is 12.92385 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkConsumeGroupsMirroringTest&test_method=test_continuous_group_sync |
| ShadowLinkingReplicationTests                  | test_auto_prefix_trimming                      | {"source_cluster_spec": {"cluster_type": "redpanda"}, "with_failures": true} | integration | https://buildkite.com/redpanda/redpanda/builds/75011#019a229b-0328-4155-aef4-b66c3d325db4 | FLAKY         | 19/21    | upstream reliability is '93.33333333333333'. current run reliability is '90.47619047619048'. drift is 2.85714 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_auto_prefix_trimming         |
| WriteCachingFailureInjectionE2ETest            | test_crash_all                                 | {"use_transactions": false}                                                  | integration | https://buildkite.com/redpanda/redpanda/builds/75011#019a229b-0327-4d9e-9959-0321323713de | FLAKY         | 18/21    | upstream reliability is '82.60061919504645'. current run reliability is '85.71428571428571'. drift is -3.11367 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all              |
| src/v/cluster/tests/shard_placement_table_test | src/v/cluster/tests/shard_placement_table_test |                                                                              | unit        | https://buildkite.com/redpanda/redpanda/builds/75011#019a2261-dc8a-4209-8972-1b54f406309c | FAIL          | 0/1      |                                                                                                                                                                         |                                                                                                                                                                               |
</details>
<!--/body-->

bashtanov committed:
raft/compaction_coordinator: tolerate attempts to move MTRO backwards

MCCO = max cleanly compacted offset of a replica
MTRO = max tombstone removal offset of the raft group

We don't allow neither MCCO nor MTRO to go backwards, and should ignore
attempts to do so. These should not be logged as errors though, see code
comments.

dotnwat commented on src/v/raft/compaction_coordinator.cc:
```diff
@@ -295,6 +297,7 @@ void compaction_coordinator::send_mtro_to_followers() {
                 })
+                .finally([holder = std::move(holder)] {})
                 .discard_result();
           });
```
curious: did you intend to hold the gate until `fstat.mtro_sender->submit` completed, rather than only this nested sleep? cc @bharathv @mmaslankaprv

bashtanov replied:
I intended to hold until the outermost lambda (i.e. the submit argument) completes (i.e. its future is ready). This is needed because `send_mtro_to_follower` uses `_client_protocol`, which is a reference to a member of `consensus` class, and the gate guards from destruction of `consensus` object.

Submit argument may complete later that sumbit itself completes if the submission is aborted. It can happen either when a new task is submitted to the same executor or when the executor is destructed.
